g3-icap-client: connection pool reuse and idle checks

DanielHaimanot opened the pull request:
1. Replaced `try_send` with `send` in order to await for completion and avoid closing new TCP connections immediately due to a full pool_sender cmd buffer channel. 
2. Avoid race condition caused by calling `CheckConnection` for each connection timeout -- connections created within the same `idle_timeout` epoch can create up-to `current_idle..min_idle` connections each
3. Changed `max_idle_count` and `min_idle_count` condition check to avoid immediately closing new connections caused by (1) and (2) -- create connections up-to min_idle only on `check_interval tick` and skip adding existing connections to the re-use pool at max_idle.  

![image](https://github.com/user-attachments/assets/0d0177d9-db6d-46e3-bfe8-53a9b522e949)

![image](https://github.com/user-attachments/assets/fdd3569a-6f69-40cb-8a66-369797dd048c)

zh-jq-b commented:
> 1. Replaced `try_send` with `send` in order to await for completion and avoid closing new TCP connections immediately due to a full pool_sender cmd buffer channel.

Good catch.

> 2. Avoid race condition caused by calling `CheckConnection` for each connection timeout -- connections created within the same `idle_timeout` epoch can create up-to `current_idle..min_idle` connections each

I prefer to use `self.pool_cmd_receiver.try_recv()`  here https://github.com/bytedance/g3/blob/037402557458747dc66fc5c31d50e2b6aacae977/lib/g3-icap-client/src/service/pool.rs#L89 to receive all pool commands at once, and then execute only 1 `IcapServicePoolCommand::CheckConnection` if found (after all other pool commands executed).

> 3. Changed `max_idle_count` and `min_idle_count` condition check to avoid immediately closing new connections caused by (1) and (2) -- create connections up-to min_idle only on `check_interval tick` and skip adding existing connections to the re-use pool at max_idle.

Will this be still needed after (1) and (2) get fixed?

DanielHaimanot commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -122,24 +123,27 @@ impl IcapServicePool {
-                        let _ = pool_sender.try_send(IcapServicePoolCommand::SaveConnection(conn));
-                    }
-                });
+        for _i in current_idle_count..min_idle_count {
```
New to using rust, it seems like this pattern is OK with the latest version; added it back regardless for clarity regardless -- [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=434b6f2cde82427dfb0abcc4ed157a51)

zh-jq-b commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -124,12 +125,14 @@ impl IcapServicePool {
-                    if let Ok(conn) = conn_creator.create().await {
-                        let _ = pool_sender.try_send(IcapServicePoolCommand::SaveConnection(conn));
-                    }
+                    let _ = pool_sender
+                        .send(IcapServicePoolCommand::CreateConnection)
+                        .await;
```
This will take more time to create connections. You can add a new common `do_create_connection()` function if you want to reuse code.

zh-jq-b commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -124,12 +125,14 @@ impl IcapServicePool {
         let min_idle_count = self.config.connection_pool.min_idle_count();
         if current_idle_count < min_idle_count {
             for _i in current_idle_count..min_idle_count {
+                if min_idle_count <= self.idle_conn_count() {
+                    break;
+                }
```
This is useless as the idle count value can only be increased in `self.save_connection()`, which won't be called before the for loop ends.

zh-jq-b replied:
Yes but there is no concurrency here, it can only be called after the for loop ends.

DanielHaimanot replied:
my bad, got it

zh-jq-b commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -157,35 +160,51 @@ impl IcapServicePool {
+        tokio::spawn(async move {
+            eof_poller.into_running(idle_timeout).await;
+            idle_count.fetch_sub(1, Ordering::Relaxed);
+            let _ = pool_sender.try_send(IcapServicePoolCommand::CreateConnection);
```
this should also be changed to `send().await`.

DanielHaimanot commented:
> > 1. Replaced `try_send` with `send` in order to await for completion and avoid closing new TCP connections immediately due to a full pool_sender cmd buffer channel.
> 
> Good catch.
> 
> > 2. Avoid race condition caused by calling `CheckConnection` for each connection timeout -- connections created within the same `idle_timeout` epoch can create up-to `current_idle..min_idle` connections each
> 
> I prefer to use `self.pool_cmd_receiver.try_recv()` here
> 
> https://github.com/bytedance/g3/blob/037402557458747dc66fc5c31d50e2b6aacae977/lib/g3-icap-client/src/service/pool.rs#L89
> 
> to receive all pool commands at once, and then execute only 1 `IcapServicePoolCommand::CheckConnection` if found (after all other pool commands executed).
> > 3. Changed `max_idle_count` and `min_idle_count` condition check to avoid immediately closing new connections caused by (1) and (2) -- create connections up-to min_idle only on `check_interval tick` and skip adding existing connections to the re-use pool at max_idle.
> 
> Will this be still needed after (1) and (2) get fixed?

Yes, I think the above will resolve the main issue; however, I still have some concern about the `idle_timeout` for certain use-case -- closed connections will stay in `TCP_WAIT` for some time after we closed it. Considering icap connections often share the same host table, it could temp exhaust port resources with modest pool params. 

I am thinking making icap pool `idle_timeout` optional (eg., disable connection idle timeout) would minimize this scenario. Wdyt?

zh-jq-b commented:
> Yes, I think the above does resolve the reuse issue; however, I still have some concern about the `idle_timeout` for certain use-case -- closed connections will stay in `TCP_WAIT` for some time after we closed it. Considering icap connections often share the same host table, it could temp exhaust port resources with modest pool params.

Unix Domain Sockets may be a better option to be used locally. You can submit a PR to add support.

> 
> I am thinking making icap pool `idle_timeout` optional (eg., disable connection idle timeout) would minimize this scenario. 

`idle_time` is a way for the client to close the connection before the server close it. The client won't known whether the server has sent a FIN at the same time it want to reuse an IDLE connection. With a **correct** `idle_time` you can make sure that the server won't close the connection when reuse.

zh-jq-b commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -157,35 +155,53 @@ impl IcapServicePool {
             IcapServicePoolCommand::SaveConnection(conn) => self.save_connection(conn),
             IcapServicePoolCommand::UpdateOptions(options) => self.options = Arc::new(options),
-            IcapServicePoolCommand::CheckConnection => self.check(),
+            IcapServicePoolCommand::CreateConnection => self.do_create(),
```
`self.create()` is still needed here as `self.do_create()` doesn't need to check idle count.

zh-jq-b commented on lib/g3-icap-client/src/service/pool.rs:
```diff
@@ -157,35 +155,53 @@ impl IcapServicePool {
-            let idle_timeout = self.config.connection_pool.idle_timeout();
+    fn do_create(&mut self) {
+        let max_idle_count = self.config.connection_pool.max_idle_count();
+        if self.idle_conn_count() < max_idle_count {
```
let's move the check here to `self.create()` and the inner to `self.do_create()`.
